### ISSUE TITLE
TypeError when deepcopying (Faked)WBEMConnection on py27/py34

### ISSUE THREAD

**Issue by andy-maier** [pywbem/pywbem#2750]
This is to address issue https://github.com/pywbem/pywbemtools/issues/1030.

The solution in the stackoverflow answer https://stackoverflow.com/a/53660367/1424462 to use a `__deepcopy__()` special method is a good solution.

Besides not copying the logging related objects that caused that issue, we also need to answer the question what else would make sense not to be copied:

For WBEMConnection:
* last request/response data?  
* recorder state?
* state of the request session with socket pools etc. that hangs off of it?

NOTE: In pywbemcli we don't use the last... data at all right now. I tried to create a command that would list it but was not happy with the result. Not sure yet about recorder state.   In pywbemcli we use the deepcopy ONLY when we are changing some parameter of the connection (see that list below) or the state of logging. In the case of change of connection parameters we close and reopen the connection because the only way to get most of those parameters into WBEMConnection is through the __init__. 

For FakedWBEMConnection:
* state of its repository?
* any state potentially maintained by its registered providers?

KS: I think that separating the repository  and possibly the state of registered providers is important.  We don't want that type of thing to go away each time we modify some parameter in the connection with a tool like pywbemcli.  Since the user providers are synced with the repository in many cases (ex. namespace instances and the namespace provider) it makes sense that the state of each needs to stay in sync.

**Comment by KSchopmeyer** [comment]
Fields that we currently allow to be modified in pywbemcli  PywbemServer in an interactive command.

These are tied to a PywbemServer. These can be modified in an interactive command but the only way to effect the change is to close and reopen the connection PywbemServer close() and the next command initiation reopens the server.  To be honest, we never really reviewed which of these options is logical before so we may want to think of some changes. 

1. server - used on connect.
2. mock_server - used on connect.
3. name - used on connect.
4. user - used on connect.  
5. password - used on connect.
6. default_namespace - used on connect.
7. timeout - set on connect but used with each request
8. use_pull - Passed to WBEMConnection in __init__
9. verify   used on connect.
10 certfile   used on connect.
11. keyfile   used on connect.
12. ca_certs   used on connect.

I seriously question if we should be allowing some of the above as part of the interactive session since they really should be driving a new PywbemServer object, and not modification of an existing one.  To me I question the parameters server, mock_server, and name at least.

pull_max_count - set on connect but used with each request

**OPTIONS SET INTO ContextObj**

In addition some parameters are only for the ContextObj that gets created for
each command so they are easy to setup

1. logging - gets set through the _logging.py module configure_from_string. Thus is impacts the connection but is not part of the connection __init__ .
2. verbose - This goes no further than pywbemcli in most commands since only a small part of pywbem has a verbose parameter (compiler)

**Comment by andy-maier** [comment]
There are different use cases for interactive mode:
* Working against a mock environment. As soon as you want to modify the repository and see the change in another command, that can only be done in interactive mode, and only when the mock environment is set at the command level (because I don't think we would want options on an interactive command to be sticky, i.e. be applied to the parent context and in place for the next interactive command). For that case, it does not really make sense to change the connection in the interactive commands.
* However, there is another use case, where the interactive mode is used just for convenience, e.g. you have a pywbemcli history independent of your terminal sessions, and there is tab expansion and help. That is a good enough reason to use the interactive mode, and you may be working against a real server. In that case, it would make sense to specify different servers or named connections to servers.

**Comment by andy-maier** [comment]
On pull_max_count: Your PywbemServer list of attributes shown above is completely in a persistent connection, plus pull_max_count. So from that perspective, pull_max_count should probably be moved into the PywbemServer object.

However, we should not start the discussion with how it is implemented, but what behavior we want to have as a result.

It does make sense for some of the connection attributes to be overridden in interactive mode. For example timeout: A user specifies a real server (directly or via named connection) at the command level and enters interactive mode. He finds that one particular command takes longer than the timeout specified at the command level, and for that case it makes sense to be able to override the timeout in interactive mode.

Another use case may be a security-sensitive user who does not like the fact that we store the password unencrypted in the persisted connection, but they still want to benefit from persisted connections. So they omit the password in the connection definition and specify it on each command (I ignore the fact for a moment that the password would also appear in the command history we persist).

And then there are corner cases where the persisted connection was not entirely correct and something needs to be fixed in interactive mode (e.g. using --no-verify when it turns out that the certificate just expired, or the cert files have been moved since the connection was defined). That would require verify, certfile, keyfile, and ca_certs to be overridable.

Another case is where a different user is used for some things. Remember, on z/OS the backend performs verification of access privileges, so another user may be needed for certain operations than the standard user stored with the connection. That would require user and password to be overridable.

The pull options are relevant only for real servers, but in the use case where interactive mode is used for real servers, it may be possible that `--use-pull no` is needed if there is some misbehavior of the server and that was not known at the time the connection was defined, or to optimize the max count or simply to play around with it to measure the effects of changing it. So the two pull options should be overridable.

For default_namespace I don't see a use case where it would need to be overridden in interactive mode, since we don't want "sticky" behavior of options (I think), and if a different namespace than the inherited default namespace is needed on an interactive command, each command can specify that without requiring the default namespace.

Bottom line: I can see use cases for all options that are currently part of a persisted connection to be overridable in interactive mode, except for the default namespace.

**Comment by andy-maier** [comment]
On overrides of connection properties: Currently, we reject overrides of connection properties on the command line: If a named connection is specified or defaulted, options that would override one of the connection properties are rejected.

I think that is not good and one should be able to have the same overrides on the command line as we would allow in interactive mode, for consistency and simplicity.

**Comment by andy-maier** [comment]
One final thought: One of the conceptual problems we have here is that we have tied the mock environment to a connection. Ideally, you would want to treat a mock environment like a server, i.e. independent of a connection to it. But I'm not sure I'm up for such a restructuring.

**Comment by KSchopmeyer** [comment]
# Table of general parameters and their requirements, usage in pywbemcli

There are two possible contexts for a parameter:

1. connection: The parameter sets a parameter of the connection and requires
   that the connection be reset to enable a changed parameter
3. cmd: The parameter modifies behavior of the command itself and does not
   impact the connection.

NOTE: There is a third possibility, in that the parameter is part of the connection
but could be set into or used by the connection without requiring that the connection be reset.

The column type defines whether the parameter applies to the WBEMConnection (real),
FakedWBEMConnection (mock) or both types.


|Context        |type | general option name | Description    | Actions               | Note |
|---------------|-----|---------------------|----------------|-----------------------|------|
| connection    |both | name                | connection name| Not implemented       |  4   |
| connection    |real | server              | url            | change url,           |      |
|               |     |                     |                | reset server. see note|      |
| connection    |real | default-namespace   |                |                       |  3   |
| connection    |real | user                | user name      |                       |  4   |
| connection    |real | password            | user password  |                       |  5   |
| connection    |both | timeout             | client timeout |                       |  1   |
| connection    |both | use-pull            |                |                       |  4   |
| connection    |both | max-pull-count      | integer        |                       |  2   |
| connection    |real | verify              | ssl verify flag|                       |  4   |
| connection    |real | certfile            |                |                       |  4   |
| connection    |real | keyfile             |                |                       |  4   |
| connection    |real | ca-certs            |                |                       |  4   |
| connection    |mock | mock-server         | scripts/mof    |                       |  4   |
| cmd           |both| verbose              |                | set in each           |      |
|               |     |                     |                | context               |      |
| cmd           |both| timestats            | Enable/disable | Set for each          |      |
|               |     |                     | statistics     | command when connected|      |
| cmd           |both| format               |                | set in each context   |      |
|               |     |                     |                | context obj, available|      |
|               |     |                     |                | for each cmd          |      |
| cmd           |both| log                  | config/enable  |Changes server         |      |
|               |     |                     | logging        | but now set for       |      |
|               |     |                     |                | eacn interactive cmd  |      |
| cmd           |both| pdb                  | enable brkpnt  | set at startup and    |      |
|               |     |                     |                | not changed           |      |
|               |     | connections-file    |                |                       |      |



Notes:

1. The WBEMConnection timeout is only settable at the level of the connection  __init__ but could be
   settable at any time the connection exists because the value is passed to the requests code each time
   a request is issued. While this applies to the connection and is persisted with the connection
   definition, thee usage could be considered either connection or session.

2. max-pull-count is maintained in the context_obj and each instance enumerate, assoc, reference, etc.
   that needs it gets it from the contextobj. Therefore in the interactive mode setting this value through
   the general options sets the max_pull-count for that command and the original value is restored for the
   next command.

3. default_namespace is settable in WBEMConnection and therefore could be considered modifiable for
   each command.  Thus, an interactive command could set it and that would be the default for that
   command.  However, since each command also has its own namespace parameter this is a duplication
   of effort.

4. This parameter can only be set into WBEMConnection on init of the connection. It is also only used
   on connection.  Modifying this in an interactive command causes the connection to be closed and
   reopened.  In the current implementation, generally this parameter causes the connection to be closed
   so the connect() with the command execution inserts the modified parameter into the connection.
   Also, the succeeding command resets the parameter back to the cmd line definition and again closes
   the connection.


# Initialization parametes for WBEMConnection and FakedWBEMConnection

WBEMConnection _init paramsurl, creds=None, default_namespace=None, x509=None,
 ca_certs=None, no_verification=False, timeout=None, use_pull_operations=False, stats_enabled=False, proxies=None

# WBEMConnection properties have the following setters:

1. default_namespace
2. operation_recorder_enabled
3. stats_enabled
4. debug

FakedWBEMConnection(default_namespace='root/cimv2', use_pull_operations=False, stats_enabled=False, timeout=None,
response_delay=None, disable_pull_operations=None, url=None

# Known issues:

max-pull-count is considered cmd level but could be session. Max pull-count is only
used at the command level. It is carried in the context and each command that requires it gets
it from the context.  Therefore it is used for the current command if that command is an
instance enumerate, associators, etc.

**Comment by andy-maier** [comment]
From your table above, I created this table that describes the desired behavior for different override situations.

Next steps:
* Agree on desired behavior
* Determine changes to existing implementation

Override columns:

* "conn -> opt" - If the `--name` option is specified on the command, what should happen when the
  general option is specified. This applies to both command line commands and interactive session
  commands.

* "cmd -> int" - If there is a connection inherited from the command line to an interactive session,
  what should happen when the option is specified on an interactive command.

| context    | type | general option name | override conn -> opt  | override cmd -> int   | gaps conn -> opt | gaps cmd -> int |
|------------|------|---------------------|-----------------------|-----------------------|------------------|-----------------|
| connection | both | name                | N/A                   | use new connection    |                  |                 |
| connection | real | server              | rejected              | use new url           |                  |                 |
| connection | real | default-namespace   | use new parm          | use new parm          |                  |                 |
| connection | real | user                | use new parm          | use new parm          |                  |                 |
| connection | real | password            | use new parm          | use new parm          |                  |                 |
| connection | both | timeout             | use new parm          | use new parm          |                  |                 |
| connection | both | use-pull            | use new parm          | use new parm          |                  |                 |
| connection | both | max-pull-count      | use new parm          | use new parm          |                  |                 |
| connection | real | verify              | use new parm          | use new parm          |                  |                 |
| connection | real | certfile            | use new parm          | use new parm          |                  |                 |
| connection | real | keyfile             | use new parm          | use new parm          |                  |                 |
| connection | real | ca-certs            | use new parm          | use new parm          |                  |                 |
| connection | mock | mock-server         | rejected              | use new mock          |                  |                 |
| cmd        | both | verbose             | use new parm          | use new parm          |                  |                 |
| cmd        | both | timestats           | use new parm          | use new parm          |                  |                 |
| cmd        | both | format              | use new parm          | use new parm          |                  |                 |
| cmd        | both | log                 | use new parm          | use new parm          |                  |                 |
| cmd        | both | pdb                 | use new parm          | use new parm          |                  |                 |

**Comment by KSchopmeyer** [comment]
Accidently hit the close issue column

**Comment by andy-maier** [comment]
PR #2751 has been updated to now add `copy()` methods to the relevant classes, instead of using `__deepcopy__()`. This addresses this issue.
